scripts/windows/test/Setup-Env.ps1

```diff
@@ -0,0 +1,112 @@
+
+    .EXAMPLE
+        .\Setup-Env.ps1
+            -E2ETestFolder "C:\Data\e2etests"
+            -ArtifactImageBuildNumber "20190101.1"
+
+    .NOTES
+        This script is to setup the environment for running E2E tests.
+        It detects and handles different Windows version differently.
+        It triggers reboot on Windows IoT Core.
+    #>
+
+[CmdletBinding()]
+Param (
+    [ValidateNotNullOrEmpty()]
+    [ValidateScript({(Test-Path $_ -PathType Container)})]
+    [string] $E2ETestFolder = ".",
+
+    [ValidateNotNullOrEmpty()]
+    [string] $ArtifactImageBuildNumber = $(Throw "Artifact image build number is required")
+)
+
+Function PrintHighlightedMessage
+{
+    param ([string] $heading)
+
+    Write-Host -f Cyan $heading
+}
+
+Set-StrictMode -Version "Latest"
+$ErrorActionPreference = "Stop"
+$global:ProgressPreference = "SilentlyContinue"
+
+
+PrintHighlightedMessage "Environment Setup"
+
+$osEdition = (Get-ItemProperty 'HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion').EditionID
+Write-Host "Configuring OS edition $osEdition..."
+
+$IoTEdgedArtifactFolder = Join-Path $E2ETestFolder "artifacts\iotedged-windows"
+$InstallationScriptPath = Join-Path $E2ETestFolder "artifacts\core-windows\scripts\windows\setup\IotEdgeSecurityDaemon.ps1"
+Invoke-Expression $InstallationScriptPath
+
+If ($osEdition -eq "IoTUAP")    # Windows IoT Core - update iotedge
+{
+	# Check iotedge version
+	$serviceName = 'iotedge'
+	
+	If (Get-Service $serviceName -ErrorAction SilentlyContinue) {
+		$serviceVersion = Invoke-Expression "$serviceName version"
+		Write-Host "Installed $serviceName service version: $serviceVersion"
+		Write-Host "Artifact $serviceName service version: $ArtifactImageBuildNumber"
+		If ((Get-Service $serviceName).Status -eq 'Running') {
+			Stop-Service $serviceName
+		}
+		
+		Write-Host "Cleanup existing containers..."
+		docker -H npipe:////./pipe/iotedge_moby_engine rm -f $(docker -H npipe:////./pipe/iotedge_moby_engine ps -aq)
+		#docker rm -f $(docker ps -aq)
+		
+		Write-Host "Attempt to update $serviceName..."
+		try {
+			# triggers reboot
+			Update-IoTEdge -ContainerOs Windows -OfflineInstallationPath $IoTEdgedArtifactFolder
+		}
+		catch {
+			$testExitCode = [String]::Format("0x{0:X}", $LastExitCode)
+			Write-Host "Update attempt unsuccessful (error code $testExitCode)."
+			if ($LastExitCode -eq -2145877235) {
+				Write-Host "A newer version is already installed on the device."
+			}
+			# triggers reboot
+			#shutdown -r -t 5
+		}
+	} Else {
+		Write-Host "$serviceName not found."
+		Write-Host "Installing $serviceName..."
+		# triggers reboot
+		deploy-iotedge
+	}
+}
+Else    # Windows Pro/Enterprise/Server - uninstall iotedge
+{
+	Write-Host "Uninstall iotedged"
+	# no reboot
+	Uninstall-IoTEdge -Force
+}
+
+# hide exit error caused by target reboot
+# next step in the pipeline will verify if the device comes back successfully
+Exit 0
```

Comment:
this should move inside "If ($osEdition -eq "IoTUAP")" block.

Reply:
Good point, moved.